File: src/test/java/YatzyTest.java
import org.junit.*;
import static org.junit.Assert.*;

public class YatzyTest {

    @Test
    public void change() {
        assertEquals(15, new Yatzy(2,3,4,5,1).chance());
        assertEquals(16, new Yatzy(3,3,4,5,1).chance());
        assertEquals(14, new Yatzy(1,1,3,3,6).chance());
        assertEquals(21, new Yatzy(4,5,5,6,1).chance());

    }

    @Test public void yatzy() {
        assertEquals(50, new Yatzy(4,4,4,4,4).yatzy());
        assertEquals(50, new Yatzy(6,6,6,6,6).yatzy());
        assertEquals(0, new Yatzy(6,6,6,6,3).yatzy());
        assertEquals(50, new Yatzy(1,1,1,1,1).yatzy());
        assertEquals(0, new Yatzy(1,1,1,2,1).yatzy());
    }

    @Test public void ones() {
        assertEquals (1, new Yatzy(1,2,3,4,5).ones());
        assertEquals(2, new Yatzy(1,2,1,4,5).ones());
        assertEquals(0,  new Yatzy(6,2,2,4,5).ones());
        assertEquals(4, new Yatzy(1,2,1,1,1).ones());
        assertEquals(0, new Yatzy(3,3,3,4,5).ones());
    }

    @Test
    public void twos() {
        assertEquals(4, new Yatzy(1,2,3,2,6).twos());
        assertEquals(10, new Yatzy(2,2,2,2,2).twos());
        assertEquals(4, new Yatzy(2,3,2,5,1).twos());

    }

    @Test
    public void threes() {
        assertEquals(6, new Yatzy(1,2,3,2,3).threes());
        assertEquals(12, new Yatzy(2,3,3,3,3).threes());
    }

    @Test
    public void fours()
    {
        assertEquals(12, new Yatzy(4,4,4,5,5).fours());
        assertEquals(8, new Yatzy(4,4,5,5,5).fours());
        assertEquals(4, new Yatzy(4,5,5,5,5).fours());
        assertEquals(8, new Yatzy(1,1,2,4,4).fours());

    }

    @Test
    public void fives() {
        assertEquals(10,  new Yatzy(4,4,4,5,5).fives());
        assertEquals(15,  new Yatzy(4,4,5,5,5).fives());
        assertEquals(20,  new Yatzy(4,5,5,5,5).fives());
    }

    @Test
    public void sixes() {
        assertEquals(0, new Yatzy(4,4,4,5,5).sixes());
        assertEquals(6, new Yatzy(4,4,6,5,5).sixes());
        assertEquals(18, new Yatzy(6,5,6,6,5).sixes());
    }

    @Test
    public void pair() {
        assertEquals(6, new Yatzy(3,4,3,5,6).pair());
        assertEquals(10, new Yatzy(5,3,3,3,5).pair());
        assertEquals(12, new Yatzy(5,3,6,6,5).pair());

        assertEquals(0, new Yatzy(1,2,3,4,5).pair());
        assertEquals(8, new Yatzy(3,3,3,4,4).pair());
        assertEquals(12,new Yatzy(1,1,6,2,6).pair());
        assertEquals(6, new Yatzy(3,3,3,4,1).pair());
        assertEquals(6, new Yatzy(3,3,3,3,1).pair());

    }

    @Test
    public void twoPairs() {
        assertEquals(16, new Yatzy(3,3,5,4,5).twoPairs());
        assertEquals(16, new Yatzy(3,3,5,5,5).twoPairs());

        assertEquals(8, new Yatzy(1,1,2,3,3).twoPairs());
        assertEquals(0, new Yatzy(1,1,2,3,4).twoPairs());
        assertEquals(6, new Yatzy(1,1,2,2,2).twoPairs());
        assertEquals(0, new Yatzy(3,3,3,3,1).twoPairs());

    }

    @Test
    public void threeOfAKind()
    {
        assertEquals(9, new Yatzy(3,3,3,4,5).threeOfAKind());
        assertEquals(15, new Yatzy(5,3,5,4,5).threeOfAKind());
        assertEquals(9, new Yatzy(3,3,3,3,5).threeOfAKind());
        assertEquals(9, new Yatzy(3,3,3,3,3).threeOfAKind());

        assertEquals(0, new Yatzy(3,3,4,5,6).threeOfAKind());
        assertEquals(9, new Yatzy(3,3,3,3,1).threeOfAKind());

    }

    @Test
    public void fourOfAKind() {
        assertEquals(12, new Yatzy(3,3,3,3,5).fourOfAKind());
        assertEquals(20, new Yatzy(5,5,5,4,5).fourOfAKind());

        assertEquals(8, new Yatzy(2,2,2,2,5).fourOfAKind());
        assertEquals(0, new Yatzy(2,2,2,5,5).fourOfAKind());
        assertEquals(8, new Yatzy(2,2,2,2,2).fourOfAKind());

    }

    @Test
    public void smallStraight() {
        assertEquals(15, new Yatzy(1,2,3,4,5).smallStraight());
        assertEquals(15, new Yatzy(2,3,4,5,1).smallStraight());
        assertEquals(0, new Yatzy(1,2,2,4,5).smallStraight());

    }

    @Test
    public void largeStraight() {
        assertEquals(20, new Yatzy(6,2,3,4,5).largeStraight());
        assertEquals(20, new Yatzy(2,3,4,5,6).largeStraight());
        assertEquals(0, new Yatzy(1,2,2,4,5).largeStraight());
    }

    @Test
    public void fullHouse() {
        assertEquals(18, new Yatzy(6,2,2,2,6).fullHouse());
        assertEquals(0, new Yatzy(2,3,4,5,6).fullHouse());

        assertEquals(8, new Yatzy(1,1,2,2,2).fullHouse());
        assertEquals(0, new Yatzy(2,2,3,3,4).fullHouse());
        assertEquals(0, new Yatzy(4,4,4,4,4).fullHouse());
    }
}

File: scala/src/test/scala/YatzyTest.scala
package org.fodehissirou

import YatzyKata.*

import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class YatzyTests extends AnyFunSuite with Matchers {


  test("Chance")  {
    assert(YatzyKata.score(List(2, 3, 4, 5, 1), Chance) == 15)
    assert(YatzyKata.score(List(3, 3, 4, 5, 1), Chance)== 16)
    assert(YatzyKata.score(List(1, 1, 3, 3, 6), Chance) == 14)
    assert(YatzyKata.score(List(4, 5, 5, 6, 1), Chance) == 21)
  }

  test("yatzy") {
    assert(YatzyKata.score(List(4, 4, 4, 4, 4), Yatzy) == 50)
    assert(YatzyKata.score(List(6, 6, 6, 6, 6), Yatzy) == 50)
    assert(YatzyKata.score(List(6, 6, 6, 6, 3), Yatzy) == 0)
    assert(YatzyKata.score(List(1, 1, 1, 1, 1), Yatzy) == 50)
    assert(YatzyKata.score(List(1, 1, 1, 2, 1), Yatzy) == 0)
  }

  test("ones") {
    assert(YatzyKata.score(List(1, 2, 3, 4, 5), Ones) == 1)
    assert(YatzyKata.score(List(1, 2, 1, 4, 5), Ones) == 2)
    assert(YatzyKata.score(List(6, 2, 2, 4, 5), Ones) == 0)
    assert(YatzyKata.score(List(1, 2, 1, 1, 1), Ones) == 4)
    assert(YatzyKata.score(List(3, 3, 3, 4, 5), Ones) == 0)
  }

  test("twos") {
    assert(YatzyKata.score(List(1, 2, 3, 2, 6), Twos) == 4)
    assert(YatzyKata.score(List(2, 2, 2, 2, 2), Twos) == 10)
    assert(YatzyKata.score(List(2, 3, 2, 5, 1), Twos) == 4)
  }

  test("Threes") {
    assert(YatzyKata.score(List(1, 2, 3, 2, 3), Threes) == 6)
    assert(YatzyKata.score(List(2, 3, 3, 3, 3), Threes) == 12)
  }

  test("fours") {
    assert(YatzyKata.score(List(4, 4, 4, 5, 5), Fours) == 12)
    assert(YatzyKata.score(List(4, 4, 5, 5, 5), Fours) == 8)
    assert(YatzyKata.score(List(4, 5, 5, 5, 5), Fours) == 4)
    assert(YatzyKata.score(List(1, 1, 2, 4, 4), Fours) == 8)
  }

  test("fives")  {
    assert(YatzyKata.score(List(4, 4, 4, 5, 5), Fives) == 10)
    assert(YatzyKata.score(List(4, 4, 5, 5, 5), Fives) == 15)
    assert(YatzyKata.score(List(4, 5, 5, 5, 5), Fives) == 20)
  }

  test("sixes") {
    assert(YatzyKata.score(List(4, 4, 4, 5, 5), Sixes) == 0)
    assert(YatzyKata.score(List(4, 4, 6, 5, 5), Sixes) == 6)
    assert(YatzyKata.score(List(6, 5, 6, 6, 5), Sixes) == 18)
  }

  test("pair") {
    assert(YatzyKata.score(List(3, 4, 3, 5, 6), Pair) == 6)
    assert(YatzyKata.score(List(5, 3, 3, 3, 5), Pair) == 10)
    assert(YatzyKata.score(List(5, 3, 6, 6, 5), Pair) == 12)
    assert(YatzyKata.score(List(1, 2, 3, 4, 5), Pair) == 0)
    assert(YatzyKata.score(List(3, 3, 3, 4, 4), Pair) == 8)
    assert(YatzyKata.score(List(1, 1, 6, 2, 6), Pair) == 12)
    assert(YatzyKata.score(List(3, 3, 3, 4, 1), Pair) == 6)
    assert(YatzyKata.score(List(3, 3, 3, 3, 1), Pair) == 6)
  }

  test("twoPairs") {
    assert(YatzyKata.score(List(3, 3, 5, 4, 5), TwoPairs) == 16)
    assert(YatzyKata.score(List(3, 3, 5, 5, 5), TwoPairs) == 16)
    assert(YatzyKata.score(List(1, 1, 2, 3, 3), TwoPairs) == 8)
    assert(YatzyKata.score(List(1, 1, 2, 3, 4), TwoPairs) == 0)
    assert(YatzyKata.score(List(1, 1, 2, 2, 2), TwoPairs) == 6)
    assert(YatzyKata.score(List(3, 3, 3, 3, 1), TwoPairs) == 0)
  }

  test("threeOfAKind") {
    assert(YatzyKata.score(List(3, 3, 3, 4, 5), ThreeOfAKind) == 9)
    assert(YatzyKata.score(List(5, 3, 5, 4, 5), ThreeOfAKind) == 15)
    assert(YatzyKata.score(List(3, 3, 3, 3, 5), ThreeOfAKind) == 9)
    assert(YatzyKata.score(List(3, 3, 3, 3, 3), ThreeOfAKind) == 9)
    assert(YatzyKata.score(List(3, 3, 4, 5, 6), ThreeOfAKind) == 0)
    assert(YatzyKata.score(List(3, 3, 3, 3, 1), ThreeOfAKind) == 9)
  }

  test("FourOfAKind") {
    assert(YatzyKata.score(List(3, 3, 3, 3, 5), FourOfAKind) == 12)
    assert(YatzyKata.score(List(5, 5, 5, 4, 5), FourOfAKind) == 20)
    assert(YatzyKata.score(List(2, 2, 2, 2, 5), FourOfAKind) == 8)
    assert(YatzyKata.score(List(2, 2, 2, 5, 5), FourOfAKind) == 0)
    assert(YatzyKata.score(List(2, 2, 2, 2, 2), FourOfAKind) == 8)
  }

  test("smallStraight"){
    assert(YatzyKata.score(List(1, 2, 3, 4, 5), SmallStraight) == 15)
    assert(YatzyKata.score(List(2, 3, 4, 5, 1), SmallStraight) == 15)
    assert(YatzyKata.score(List(1, 2, 2, 4, 5), SmallStraight) == 0)
  }

  test("largeStraight"){
    assert(YatzyKata.score(List(6, 2, 3, 4, 5), LargeStraight) == 20)
    assert(YatzyKata.score(List(2, 3, 4, 5, 6), LargeStraight) == 20)
    assert(YatzyKata.score(List(1, 2, 2, 4, 5), LargeStraight) == 0)
  }

  test("fullHouse"){
    assert(YatzyKata.score(List(6, 2, 2, 2, 6), FullHouse) == 18)
    assert(YatzyKata.score(List(2, 3, 4, 5, 6), FullHouse) == 0)

    assert(YatzyKata.score(List(1, 1, 2, 2, 2), FullHouse) == 8)
    assert(YatzyKata.score(List(2, 2, 3, 3, 4), FullHouse) == 0)
    assert(YatzyKata.score(List(4, 4, 4, 4, 4), FullHouse) == 0)
  }

}